src/bun.js/bindings/JSStringDecoder.cpp

```diff
@@ -0,0 +1,383 @@
+            return JSC::jsString(vm, WTF::String(u"\uFFFD", 1));
+        }
+        if (m_lastNeed > 1 && length > 1) {
+            if ((bufPtr[1] & 0xC0) != 0x80) {
+                m_lastNeed = 1;
+                return JSC::jsString(vm, WTF::String(u"\uFFFD", 1));
+            }
+            if (m_lastNeed > 2 && length > 2) {
+                 if ((bufPtr[2] & 0xC0) != 0x80) {
+                    m_lastNeed = 2;
+                    return JSC::jsString(vm, WTF::String(u"\uFFFD", 1));
+                }
+            }
+        }
+    }
+
+    if (m_lastNeed <= length) {
+        memmove(m_lastChar + m_lastTotal - m_lastNeed, bufPtr, m_lastNeed);
+        return JSC::JSValue::decode(Bun__encoding__toString(m_lastChar, m_lastTotal, globalObject, static_cast<uint8_t>(m_encoding)));
+    }
+    memmove(m_lastChar + m_lastTotal - m_lastNeed, bufPtr, length);
+    m_lastNeed -= length;
+    return JSC::jsEmptyString(vm);
+}
+
+// Checks the type of a UTF-8 byte, whether it's ASCII, a leading byte, or a
+// continuation byte. If an invalid byte is detected, -2 is returned.
+int8_t utf8CheckByte(uint8_t byte) {
+    if (byte <= 0x7F) return 0;
+    else if ((byte >> 5) == 0x06) return 2;
+    else if ((byte >> 4) == 0x0E) return 3;
+    else if ((byte >> 3) == 0x1E) return 4;
+    return (byte >> 6) == 0x02 ? -1 : -2;
+}
+
+// Checks at most 3 bytes at the end of a Buffer in order to detect an
+// incomplete multi-byte UTF-8 character. The total number of bytes (2, 3, or 4)
+// needed to complete the UTF-8 character (if applicable) are returned.
+uint8_t JSStringDecoder::utf8CheckIncomplete(JSC::JSUint8Array* buf, uint32_t i)
+{
+    uint32_t j = buf->length() - 1;
+    uint8_t* bufPtr = buf->typedVector();
+    if (j < i) return 0;
+    int8_t nb = utf8CheckByte(bufPtr[j]);
+    if (nb >= 0) {
+        if (nb > 0) m_lastNeed = nb - 1;
+        return nb;
+    }
+    if (--j < i || nb == -2) return 0;
+    nb = utf8CheckByte(bufPtr[j]);
+    if (nb >= 0) {
+        if (nb > 0) m_lastNeed = nb - 2;
+        return nb;
+    }
+    if (--j < i || nb == -2) return 0;
+    nb = utf8CheckByte(bufPtr[j]);
+    if (nb >= 0) {
+        if (nb > 0) {
+            if (nb == 2) nb = 0;else m_lastNeed = nb - 3;
+        }
+        return nb;
+    }
+    return 0;
+}
+
+JSC::JSValue JSStringDecoder::text(JSC::VM& vm, JSC::JSGlobalObject* globalObject, JSC::JSUint8Array* buf, uint32_t offset)
+{
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+    uint32_t length = buf->length();
+    uint8_t* bufPtr = buf->typedVector();
+
+    switch (m_encoding) {
+    case BufferEncodingType::ucs2:
+    case BufferEncodingType::utf16le: {
+        if ((length - offset) % 2 == 0) {
+            JSString* r = JSC::JSValue::decode(Bun__encoding__toString(bufPtr + offset, length - offset, globalObject, static_cast<uint8_t>(m_encoding))).toString(globalObject);
+            RETURN_IF_EXCEPTION(throwScope, JSC::jsUndefined());
+            if (r->length()) {
```

Comment:
It's not good that we are creating the JSString and then reading the value from it

This is an extra string 1-2 allocations

Reply:
fixed.